Encryption.md

```diff
@@ -0,0 +1,184 @@
+6. Work with all compression and encoding mechanisms supported in Parquet.
+7. Support multiple encryption algorithms, to account for different security and 
+performance requirements.
+
+
+## Non-Goals
+1. Key management. 
+Keys, arbitrary key metadata and key retrieval callbacks are provided to Parquet API as input 
+parameters. Key storage, DEK encryption with KEK (data- and key-encryption keys), re-keying 
+etc, are out of scope and should be done above Parquet. Parquet accepts explicit keys, and 
+also provides tools to store key metadata inside the file (upon encryption), and to call 
+key retrieval callbacks with this metadata (upon decryption).
+
+
+
+## Technical Approach
+
+Each Parquet module (footer, page headers, pages, column indexes, column metadata) is 
+encrypted separately. Then it is possible to fetch and decrypt the footer, find the 
+offset of a required page, fetch it and decrypt the data. In this document, the term 
+“footer” always refers to the regular Parquet footer - the `FileMetaData` structure, and 
+its nested fields (row groups / column chunks).
+
+The results of compression of column pages are encrypted, before being written to the 
+output stream. A new Thrift structure, with a column crypto metadata, is added to 
+column chunks of the encrypted columns. This metadata provides information about the 
+column encryption keys.
+
+The results of Thrift serialization of metadata structures are encrypted, before being 
+written to the output stream.
+
+A new Thrift structure, with a file crypto metadata,  is written after the (encrypted) 
+Parquet footer, at the end of the file. This structure is not encrypted. It provides 
+information about the footer encryption key, the algorithm used to encrypt the file, etc.
+
+## Encryption algorithms
+
+Parquet encryption algorithms are based on the standard AES ciphers for symmetric encryption. 
+AES is supported in Intel and other CPUs with hardware acceleration of 
+crypto operations (“AES-NI”) - that can be leveraged by e.g. Java programs (automatically 
+via HotSpot), or C++ programs (via EVP-* functions in OpenSSL).
+
+Initially, two algorithms are implemented, one based on an AES-GCM cipher, and the other on a 
+combination of AES-GCM and AES-CTR ciphers.
+
+AES-GCM is an authenticated encryption. Besides the data confidentiality (encryption), it 
+supports two levels of integrity verification / authentication: of the data (default), and 
+of the data combined with an optional AAD (“additional authenticated data”). The default 
+authentication allows to make sure the data has not been tampered with. An AAD is a free 
+text to be signed, together with the data. The user can, for example, pass the file name 
+with its version (or creation timestamp) as the AAD, to verify the file has not been 
+replaced with an older version.
+
+Sometimes, a hardware acceleration of AES is unavialable (e.g. in Java 8). Then AES crypto 
```

Comment:
nit: typo "unavialable"

Reply:
thanks, I'll fix it. Btw, if you English is your native tongue - I'd really appreciate a language review of this doc!!